### PR TITLE
refactor(*): compile to webapp!

### PR DESCRIPTION
**Pull request by ramfox**
NEW COMMANDS:

- `yarn dev-webapp` - launches a dev server w/ hot module reload! relies on `webpack.config.webapp.dev.js`

- `yarn build-webapp` - compiles the webapp into 3 files: `index.html`, `bundle.js`, `bundle.css`. These three files need to be served in order to view them over HTTP. Relies on `webpack.config.webapp.js`

- `yarn webapp` - calls `build-webapp` under the hood, as well as `serve -l app/webapp/dist`, builds and serves the webapp

To use this webapp with your local Qri node you must add `http://localhost:5000` to your list of `API.allowedorigins`

Step 2 in #663

### CODE REVIEW

File: app/store/configureStore.webapp.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { createStore, applyMiddleware } from 'redux'
+import thunk from 'redux-thunk'
+import { createBrowserHistory } from 'history'
```
Comment (b5) [inline app/store/configureStore.webapp.ts:3]: nice catch using the browser history in production. seeing a hash-router in `yarn dev-webapp` was tripping me up until I realized that electron & `dev-webapp` share the same development store configuration

File: webpack.config.development.js [hunk 2]
```diff
@@ -52,85 +51,13 @@ module.exports = merge(baseConfig, {
...
-          options: {
-            limit: 10000,
-            mimetype: 'image/svg+xml'
-          }
-        }
-      },
-      // Common Image Formats
-      {
-        test: /\.(?:ico|gif|png|jpg|jpeg|webp)$/,
-        use: 'url-loader'
```
Comment (b5) [inline webpack.config.development.js:133]: very nice cleanup

File: app/components/App.tsx [hunk 3]
```diff
@@ -2,15 +2,15 @@ import React from 'react'
 import { Action } from 'redux'
 import { CSSTransition } from 'react-transition-group'
 import { ConnectedRouter, push } from 'connected-react-router'
-import fs from 'fs'
+
```
Comment (b5) [inline app/components/App.tsx:5]: unnecessary extra line here

File: app/components/BodyTablePreview.tsx [hunk 4]
```diff
@@ -33,9 +34,9 @@ const BodyTablePreview: React.FunctionComponent<BodyPreviewProps> = ({ data }) =
     )
   })
 
-  let headers: Array<{ type: string, title: string}> = []
+  let headers: ColumnProperties[] = []
   if (structure && structure.schema) {
-    structure.schema.items.items.forEach((column: { type: string, title: string}) => {
+    schemaColumns(structure.schema).forEach((column: ColumnProperties) => {
```
Comment (b5) [inline app/components/BodyTablePreview.tsx:39]: much nicer

File: app/components/compare/platformSpecific/CompareSidebar.electron.ts [hunk 5]
```diff
@@ -1,5 +1,21 @@
 import { remote, OpenDialogSyncOptions } from 'electron'
+import fs from 'fs'
```
Comment (b5) [inline app/components/compare/platformSpecific/CompareSidebar.electron.ts:2]: let's rename this file (and all related platformSpecific files) to `chooseFile.electron.ts`, this code is now abstracted away from `CompareSidebar`

File: app/components/compare/platformSpecific/CompareSidebar.electron.ts [hunk 6]
```diff
@@ -1,5 +1,21 @@
...
 
-export function showOpenDialogSync (opts: OpenDialogSyncOptions): string[] | undefined {
-  return remote.dialog.showOpenDialogSync(remote.getCurrentWindow(), opts)
+export function chooseFile (opts: OpenDialogSyncOptions): string[] {
```
Comment (b5) [inline app/components/compare/platformSpecific/CompareSidebar.electron.ts:4]: for stronger type safety here use a tuple return value:

```suggestion
export function chooseFile (opts: OpenDialogSyncOptions): [string, string] {
```

This'll specify that the array length _must_ be 2

File: app/components/collection/ComponentRouter.tsx [hunk 7]
```diff
@@ -94,7 +94,7 @@ export const ComponentRouterComponent: React.FunctionComponent<ComponentRouterPr
     const ext = path.extname(e.dataTransfer.files[0].path)
     if (!(ext === '.csv' || ext === '.json')) {
       // open toast for 1 second
-      openToast(ToastTypes.error, 'drag-drop', 'unsupported file format: only json and csv supported')
+      openToast('error', 'drag-drop', 'unsupported file format: only json and csv supported')
```
Comment (b5) [inline app/components/collection/ComponentRouter.tsx:97]: y no `ToastTypes.error`?

File: app/components/compare/platformSpecific/CompareSidebar.web.ts [hunk 8]
```diff
@@ -1,4 +1,4 @@
-export function showOpenDialogSync (): undefined {
-  console.log("showOpenDialogSync is not implemented for the webapp version of the Qri Desktop")
-  return undefined
+export function chooseFile (): string[] {
```
Comment (b5) [inline app/components/compare/platformSpecific/CompareSidebar.web.ts:1]: use a tuple of strings as the return value type: `[string, string]`

File: app/models/dataset.ts [hunk 9]
```diff
@@ -77,6 +77,12 @@ export interface Structure {
   schema?: Schema
 }
 
+export interface ColumnProperties {
+  title?: string
+  type?: string
```
Comment (b5) [inline app/models/dataset.ts:82]: type can be either a string or array of strings:
```suggestion
  type?: string | string[]
```
We have users that create datasets with multiple column types. Yes, lots of code will break with this assumption, but at least capturing this nuance in the type system is a start

File: app/utils/schemaColumns.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+export function schemaColumns (schema: Schema): ColumnProperties[] {
+  if (schema && schema.items && isJSONSchema7(schema.items) && schema.items.items && isJSONSchema7(schema.items.items) && Array.isArray(schema.items.items)) {
+    return schema.items.items
+  }
+  return []
+}
+
+function isJSONSchema7 (x: any): x is JSONSchema7 {
+  return (x as JSONSchema7).type !== undefined
+}
```
Comment (b5) [inline app/utils/schemaColumns.ts:13]: This is starting to look a lot like the [`tabular`](https://github.com/qri-io/dataset/blob/master/tabular/tabular.go) package in go land. Feel free to use that as inspiration, but long term it'd be great to start thinking about moving stuff like this into a `tabular` package in `qri-js` 😄.

File: test/e2e/e2e.spec.ts [hunk 11]
```diff
@@ -6,7 +6,8 @@ import TestTempRegistry from '../utils/testTempRegistry'
 import fakeDialog from 'spectron-fake-dialog'
 import { E2ETestUtils, newE2ETestUtils } from '../utils/e2eTestUtils'
 import http from 'http'
-import Dataset, { Commit, Meta, Structure } from '../../app/models/dataset'
+import Dataset, { Commit, Meta, Schema, Structure } from '../../app/models/dataset'
```
Comment (b5) [inline test/e2e/e2e.spec.ts:9]: delighted to see you're maintaining the `e2e` tests. Getting a version stood up for the web app will be a great project in 2021

**Review by b5 — APPROVED** [review]
Really really great stuff. This PR found a bunch of bugs on the backend API that I'm going to fix. Love to see this get merged

File: app/store/configureStore.webapp.ts (hunk 1 above)
Comment (ramfox) [reply app/store/configureStore.webapp.ts:3]: hmmm... this actually isn't working the way we are expecting. Even in the prod versions of `webapp` and `remote` we are still seeing the #. I need to do a deeper dive, maybe something has changed in react router dom that requires some other changes.

File: app/components/collection/ComponentRouter.tsx (hunk 7 above)
Comment (ramfox) [reply app/components/collection/ComponentRouter.tsx:97]: `ToastTypes` doesn't exist! We have a `ToastType` type that can be one of 'success', 'info', 'error'

**Pull request merged**